### PR TITLE
PyOpenCL implementation of SART method

### PR DESCRIPTION
**Pull request by vsnever**
An OpenCL implementation of SART solver. Works much faster than the CPU version if the number of light sources >10 000.

### CODE REVIEW

**Comment by vsnever** [comment]
Note that pyopencl is required to run this version of SART solver as well as the OpenCL-compatible drivers.

**Comment by mattngc** [comment]
Thanks Vlad, this is a really nice contribution. I'm looking forward to having it in the code base.

I'm going to suggest myself and @jacklovell as the merge reviewers. Jack is a bolometry expert and has done a lot of work using the inversions, so an appropriate reviewer.

File: cherab/tools/inversions/opencl/sart_opencl.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
+# -*- coding: utf-8 -*-
+# Created by Vladislav Neverov (NRC "Kurchatov Institute") for CHERAB Spectroscopy Modelling Framework
+from __future__ import print_function
+import os
+import numpy as np
+import pyopencl as cl
```
Comment (mattngc) [inline cherab/tools/inversions/opencl/sart_opencl.py]: Similarly here, can you make this a try/except import only called when this class is instantiated? Keep the reference to the module on the class maybe.

File: cherab/tools/inversions/opencl/sart_opencl.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+                                    self.geometry_matrix_device, self.cell_ray_densities_device, self.ray_lengths_device,
+                                    self.y_hat_device, self.detectors_device, self.solution_device, self.grad_penalty_device,
+                                    np.float32(relaxation), np.uint32(self.n_sources), np.uint32(self.m_detectors))
+        if self.use_atomic:
+            self.cl_prog.zero_negative(queue, self.global_work_size['trivial_sources'], self.local_work_size['default'],
+                                       self.solution_device, np.uint32(self.n_sources))
+
+
+if __name__ == "__main__":
+    pass
```
Comment (mattngc) [inline cherab/tools/inversions/opencl/sart_opencl.py]: This code is unnecessary.

File: cherab/tools/inversions/opencl/opencl_utils.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+        return get_best_gpu([platform], device_type, verbose)
+    warnings.warn('System has only %d OpenCL platforms\n' % n_platforms)
+    if non_gpu_device:
+        return get_first_device(platforms, device_type, verbose)
+
+    return get_best_gpu(platforms, device_type, verbose)
+
+
+if __name__ == "__main__":
+    device_select(platfrom_id=None, device_id=None, verbose=True)
```
Comment (mattngc) [inline cherab/tools/inversions/opencl/opencl_utils.py]: Is this necessary? I would have thought this was just for your testing and never actually gets called when SartOpencl() is used as the main entry point.

File: cherab/tools/inversions/opencl/opencl_utils.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
+# -*- coding: utf-8 -*-
+# Created by Vladislav Neverov (NRC "Kurchatov Institute") for CHERAB Spectroscopy Modelling Framework
+from __future__ import print_function
+import warnings
+import pyopencl as cl
```
[the remark this answers is not in the document]
Comment (vsnever) [reply cherab/tools/inversions/opencl/opencl_utils.py]: Done. However I decided to raise a runtime error when the function/class which depends on pyopencl is called/initialised, not when it's imported.

File: cherab/tools/inversions/opencl/sart_opencl.py (hunk 1 above)
Comment (vsnever) [reply cherab/tools/inversions/opencl/sart_opencl.py]: Done.

File: cherab/tools/inversions/opencl/sart_opencl.py (hunk 2 above)
Comment (vsnever) [reply cherab/tools/inversions/opencl/sart_opencl.py]: This code is removed.

File: cherab/tools/inversions/opencl/opencl_utils.py (hunk 3 above)
Comment (vsnever) [reply cherab/tools/inversions/opencl/opencl_utils.py]: This code is removed.

**Comment by mattngc** [comment]
Thanks for making these changes Vlad. Now waiting to see the review from @jacklovell.

File: cherab/tools/inversions/opencl/__init__.py [hunk 5]
```diff
@@ -0,0 +1,3 @@
+# -*- coding: utf-8 -*-
+from .sart_opencl import SartOpencl
```
[the remark this answers is not in the document]
Comment (vsnever) [reply cherab/tools/inversions/opencl/__init__.py]: Done.

File: cherab/tools/inversions/opencl/sart_opencl.py [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+from .opencl_utils import device_select
+try:
+    import pyopencl as cl
+except ImportError:
+    _has_pyopencl = False
+else:
+    _has_pyopencl = True
+
+
+class SartOpencl:
```
[the remark this answers is not in the document]
Comment (vsnever) [reply cherab/tools/inversions/opencl/sart_opencl.py]: Done.

File: cherab/tools/inversions/opencl/sart_opencl.py [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+
+    def __init__(self, geometry_matrix, laplacian_matrix=None, device=None, block_size=256, copy_column_major=True, block_size_row_maj=64,
+                 use_atomic=True, steps_per_thread=64):
+        """
+        This is a GPU version of invert_sart() and invert_constrained_sart() functions from the CHERAB.
+        See the documentaion to these functions here: https://cherab.github.io/documentation/tools/tomography.html.
```
[the remark this answers is not in the document]
Comment (vsnever) [reply cherab/tools/inversions/opencl/sart_opencl.py]: I changed this line.

**Comment by vsnever** [comment]
Thank you, Jack. Your suggestions are very helpful. 
I added some unit tests for `SartOpencl`. The data in `cherab/tools/tests/data` can be used to test other inversion methods as well. Note that the `SartOpencl` tests will be skipped if the `pyopencl` module is not installed. I think, it's better not to include `pyopencl` in `travis.yml` because the installation of `pyopencl` is not always straightforward. Also, the GPU (and usually the proprietary driver) is also required to successfully run `SartOpencl`.
The `verbose` parameter is set to `False` now by default, however the name of selected device will be displayed even if `verbose=False`.
A copyright notice is added to the new files.

File: cherab/tools/inversions/opencl/sart_opencl.py [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+        steps_per_thread = min(block_size, steps_per_thread)
+        steps_per_thread_row_maj = min(block_size_row_maj, steps_per_thread)
+
+        # creating OpenCL context
+        self.cl_context = cl.Context([device])
+
+        # reading and compiling OpenCL kernels
+        kernels_filename = 'sart_kernels_atomic.cl' if use_atomic else 'sart_kernels.cl'
+        kernel_source_file = os.path.join(os.path.dirname(os.path.realpath(__file__)), kernels_filename)
+        kernel_source = open(kernel_source_file).read()
```
Comment (jacklovell) [inline cherab/tools/inversions/opencl/sart_opencl.py]: This gives me a ResourceWarning when running the OpenCL unit tests:
```
/home/jlovell/FUN/cherab/cherab_core/cherab/tools/inversions/opencl/sart_opencl.py:85: ResourceWarning: unclosed file <_io.TextIOWrapper name='/home/jlovell/FUN/cherab/cherab_core/cherab/tools/inversions/opencl/sart_kernels.cl' mode='r' encoding='UTF-8'>
  kernel_source = open(kernel_source_file).read()
ResourceWarning: Enable tracemalloc to get the object allocation traceback
```
Could you please properly close the file by wrapping in a with statement:
```python
with open(kernel_source_file, "r") as f:
    kernel_source = f.read()
```

Comment (vsnever) [reply cherab/tools/inversions/opencl/sart_opencl.py]: Thanks. I fixed this.

**Pull request merged**